File: 44-wildcard-matching/wildcard-matching.js
/**
 * @param {string} s
 * @param {string} p
 * @return {boolean}
 */
var isMatch = function(s, p) {
    let l1 = s.length, l2 = p.length;
    let dp = Array.from({ length: l1 }, () => new Array(l2).fill(-1));

    const sl = (idx1, idx2) => {
        if(idx1>=l1 && idx2>=l2){
            return true;
        }
        if(idx1<l1 && idx2<l2 && dp[idx1][idx2] != -1) return dp[idx1][idx2];

        if((l1 - idx1) == 0 && p[idx2] != "*") return false;
        if((l2 - idx2) == 0 && s[idx1] != "*") return false;

        
        if(s[idx1] === "*" || p[idx2] === "*"){
            let t = false;
            t |= (idx1<l1) ? sl(idx1+1, idx2): false;
            t |= (idx2<l2) ? sl(idx1, idx2+1): false;
            t |= (idx1<l1 && idx2<l2) ? sl(idx1+1, idx2+1): false;
            if(idx1<l1 && idx2<l2) dp[idx1][idx2] = t;
            return t;
        }

        if(
            s[idx1] === p[idx2] ||
            (s[idx1] == "?" && p[idx2] != "") ||
            (p[idx2] == "?" && s[idx1] != "")){
                dp[idx1][idx2] = sl(idx1+1, idx2+1);
            return dp[idx1][idx2];
        }

        if(idx1<l1 && idx2<l2) dp[idx1][idx2] = false;

        return false;
    }


    return sl(0,0);
};

File: 44-wildcard-matching/wildcard-matching.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        ln1, ln2 = len(s), len(p)
        if ln1 == ln2 == 0:
            return True
        if ln1 == 0:
            chrc = list(set(list(p)))
            return len(chrc) == 1 and chrc[0] == "*"

        dp = {}

        def backtrack(idx1, idx2):
            store = (idx1, idx2)
            if store in dp:
                return dp[store]
            if idx1 == ln1 and idx2 == ln2:
                return True
            if idx2 == ln2-1 and p[idx2] == "*":
                return True
            if idx2 >= ln2:
                return False
            if idx1 < ln1 and s[idx1] != p[idx2] and p[idx2] != "*" and p[idx2] != "?":
                return False

            dp[store] = False
            if idx1<ln1 and (s[idx1] == p[idx2] or p[idx2] == "?"):
                dp[store] = backtrack(idx1+1, idx2+1)
                return dp[store]
            
            for i in range(idx1, ln1+1):
                while idx2<ln2 and p[idx2] == "*":
                    idx2+=1
                dp[store] |= backtrack(i,idx2)
            return dp[store]


        return backtrack(0,0)
        